### ISSUE TITLE
Have "true" bezier splines on spheres 

### ISSUE THREAD

**Issue by KristofKoch** [GenericMappingTools/gmt#3058]
**Description of the desired feature**

Have GMT support "true" bezier splines on the surface of a sphere/geoid. 

When plotting lines/polygons it might be desirable to have them plotted as bezier splines. This is already supported with `-W+s` but it works correctly only on cartesian projections as it is not a GMT function but mainly done in PostScript. This brings some disadvantages when trying to plot  ship tracks or meteorological fronts/features on a map. If implemented, it might also allow for some options to be available on bezier splines as well for example `-S~`.

Below is an example how in currently works using de11caf. Note the left vertical part where the red bezier spline should be on the left side of the green great circle line if plotted on a sphere/geoid. 

```
#!/usr/bin/env bash

cat > path.txt << END
 30  -15
 30   45
-30   45
-30  -15
END

gmt begin earth_bezier_spline
  gmt basemap -Rg -JG30/50/15c -Bpg15d -Bsg30d --MAP_GRID_PEN_SECONDARY=1p,black
  gmt plot path.txt -W2p,green
  gmt plot path.txt -W2p,red+s -A
  gmt plot path.txt -Skcrosshair/0.5c -Wthinnest,black
gmt end show
```

![2D-bezier-splines](https://forum.generic-mapping-tools.org/uploads/default/original/1X/e4996aeacb49747d5c5a7e781319cb03de96ce0a.jpeg)

**Comment by KristofKoch** [comment]
This feature request stems [from this forum post](https://forum.generic-mapping-tools.org/t/s-for-spline-doesnt-work/310).

Any new thoughts on the feasibility by the devs? If implemented could make good use of this feature.

**Comment by KristofKoch** [comment]
Awesome work @joa-quim! My first use case is/was Significant Weather Charts in Aviation:

![](https://aviationweather.gov/data/products/swh/18_sigwx_hi_b1.gif)
[Source](https://aviationweather.gov/graphics/)

Second use case is plotting ADS-B position reports from aircraft. Splines help smoothing the track visualisation between sparse position reports especially in turns. 

I haven't tried your implementation myself yet. There seems to be a problem with the lower left end of the red Bézier curve (-30 -15) in your example. It seems to extend south of S15°? I expect it to terminate on the S15° parallel.

**Comment by joa-quim** [comment]
> There seems to be a problem with the lower left end of the red Bézier curve

Good catch, and also a good lead for the next. But this particular first. It was a mistake in in the myriad of command calls I have in my history. It should have been ``Db = bezier(D1);`` instead of calling a different beast `D` (in my REPL history). The right plot is now like it should.
![bez](https://github.com/user-attachments/assets/b3d1fa52-d88b-4cc3-9982-a2957c36b25d)

This was a a _good_ mistake as it brings the issue of how we set a Bezier curve in a CLI environment.  Contrary to graphical interfaces where we can adjust the control points with the mouse, on CLI we must set them by coordinates. But that (IMO) is horribly boring/complicated. Remind that that for a cubic Bezier we must provide 4 control points, of which the result curve passes only on the end ones. Knots 2 & 3 are used to create the curve but that curve lies elsewhere in the plane. I find that to tortuous and fortunately others before me have found that as well, so I found a way to recompute the 2nd & 3rd  control points (or, knots) such that we can make the curve pass through all 4 provided knots. 

Perfect, clever guy, I thought. But the problem is that this only works with curves defined by 4 knots (the cubic Beziers): But what about if we want to have curves that pass through 5 or more points? Then, the _pass-through-control-points_ trick no longer works and I had to invent a composition curve strategy (that's what I posted by mistake in my first example). It's far from perfect, as it doesn't insure that on the gluing nodes (between curves) the first and second derivatives are equal, which is what insures a perfect smooth transition, but like that brand (f... falsely) claims, it works.

You are welcome to try this with that long curvy line around the _Global South_.

For reference and to see if the docs makes any sense to you, here is the relevant part of the on-line help of that `bezier` fun.

```
  Create a Bezier curve from a set of control points.

  A function for cubic Bezier interpolation for a given set of control points (knots). Each control point can exist in an
  N-dimensional space but only tested 2D and 3D cases. Data can be geographic or Cartesian. The default for this function is to
  compute a curve that passes through the control points. This can be changed by setting pure to true and then we instead
  compute a curve that passes through the first and last.

  When more than 4 control points are provided, we make a curve composition by connecting the first curve, made with points
  1,2,3,4, second curve, made with points 2,3,4,5 and so on. The way this connection is made is controlled by the firstcurve
  argument. With more than 4 control points, the pure option is not allowed.

  Args
  ––––

    •  D: A GMTdataset object with a matrix of control points. This matrix may have 2 (2D) or 3 columns (3D). If D
       represents geographical data, we create a spherical Bezier curve by using an intermediate step of converting the
       data to ECEF, compute the 3D curve, and then convert it back to geographic coordinates.

  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

    •  p::Matrix{<:Real}: Matrix with at least 4 control points. Same as the D case, except for the possibility of
       geographical data.

  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

    •  p1, p2, p3, p4: Vectors of control points. They can be 2D or 3D (two or three elements).

  ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────

  Kwargs
  ––––––

    •  t: Vector of values of the t paramter at which bezier curve is evaluated. By default is evaluated at 101 evenly
       spaced values between 0 and 1, but see np below.

    •  np: Number of values of the t paramter at which bezier curve is evaluated. Default is 101.

    •  pure: By default we compute a curve that passes through the control points. If pure is true, we instead compute a
       curve that passes through the first and last control points and the other two are used to control the path. This is
       what pure cubic bezier curves do.

    •  firstcurve: Applies only when the number of control points is > 4 and sets how the different curves are connected.
       If true, the first curve takes precedence. Using as example the case of two curves given by 5 control points; the
       final curve is obtained by connecting the first curve (made with the first 4 control points) with the second curve
       (made with the last 4 control points). But since these two curves shared the second, third and fourth control
       points, that part is removed from the second curve, leaving only the 4rth to 5th contribution. When firstcurve is
       false, the opposite happens. The first curve contributes only with the part from first to second control points.
```

**Comment by KristofKoch** [comment]
Lets add some more control points and circle a pole and see what is going to happen …

```
D1 = mat2ds([30 -15; 30 45; -45 60; -120 70; 180 70; 120 70; 45 60; -30 45; -30 -15], proj4="geog");
Db = bezier(D1);
plot(D1, lc=:green, lw=2, R=:g, J="G30/50/15c", B="pg15d sg30d", par=(MAP_GRID_PEN_SECONDARY="1p,black",))
plot!(Db, lc=:red, lt=2, name=:bez, show=1)
```
![bez](https://github.com/user-attachments/assets/934a0589-da59-421d-9af7-d2cbe93502c4)

I found some unexpected kinks in the line where I expected a smooth transition according to `firstcurve` but I'm not sure if I fully understand the documentation.

**Comment by joa-quim** [comment]
Those kinks are **not unexpected**.

> but I'm not sure if I fully understand the documentation.

Yes, not easy to explain. Let's try with this example.

```
D1 = bezier(mat2ds([30. -15; 30 45; -30 45; -30 -15], proj4="geog"));
D2 = bezier(mat2ds([30 45; -30 45; -30 -15; -20 0], proj4="geog"));
viz(D1, lc=:green, plot=(data=D2, lc=:red), name=:two_curves)
```

![two_curves](https://github.com/user-attachments/assets/77eaf05e-adfa-43d4-b7be-93cea1145bda)

Now, how should we connect the two curves that have **three** control points in common?

Either, privilege the first curve (the green one) using 3/4 of its length plus the last 1/4 of the second
```
 D3 = bezier(mat2ds([30. -15; 30 45; -30 45; -30 -15; -20 0], proj4="geog"));
viz(D3, name=:first)
```

![first](https://github.com/user-attachments/assets/f16ba2ee-2e30-4cb7-9785-ed4a4ecd19d0)


or privilege the second curve (the red one), using 1/4 of the first and 3/4 of the second.

```
D4 = bezier(mat2ds([30. -15; 30 45; -30 45; -30 -15; -20 0], proj4="geog"), firstcurve=false);
viz(D4, name=:second)
```

![second](https://github.com/user-attachments/assets/0f0410f6-ce68-4e1a-9d11-046f6690894f)

Given the nature of the Bezier curves, the connection may very well insert kinks

I don't think we can escape this decision. What could, eventually, be done is to find how to change the, under hood, modified control points such that at the connection node the first and second derivatives are equal so that no kinks are observed. And that is, I'm afraid, beyond my current motivation/ability to find out how.
